crates/revm/src/db/ethersdb.rs

```diff
@@ -2,14 +2,14 @@ use std::sync::Arc;
 
 use ethers_core::types::{Block, BlockId, TxHash, H160 as eH160, H256, U64 as eU64};
 use ethers_providers::Middleware;
-use tokio::runtime::Handle;
+use tokio::runtime::{Handle, Runtime};
 
 use crate::primitives::{AccountInfo, Address, Bytecode, B256, U256};
 use crate::{Database, DatabaseRef};
 
 use super::utils::HandleOrRuntime;
 
-#[derive(Debug, Clone)]
```

Comment:
@wtdcode would you explain why Clone has been removed? I'm slightly new in Rust and want to learn the technical concept behind this decision

Reply:
Because runtime doesn’t implement Clone and it doesn’t make too much sense cloning the db. Or, you could wrap it with an Rc or Arc.
What’s your use case?

Reply:
thanks
In my case, I define an ethersDB instance and by using it, I define a cacheDB. I add some base data (deploy contract, change some slots, etc) to cacheDB and after that, by cloning it, I send the same copies to multiple threads. 

For now, I get an error (when cloning cacheDB) complains about ethersDB (as internal object of cacheDB) cannot be cloned. 

because inside separate threads, I need to write to db, I tested not cloning and use Arc/Mutex. but the performance was worse than cloning.

Reply:
> thanks In my case, I define an ethersDB instance and by using it, I define a cacheDB. I add some base data (deploy contract, change some slots, etc) to cacheDB and after that, by cloning it, I send the same copies to multiple threads.
> 
> For now, I get an error (when cloning cacheDB) complains about ethersDB (as internal object of cacheDB) cannot be cloned.
> 
> because inside separate threads, I need to write to db, I tested not cloning and use Arc/Mutex. but the performance was worse than cloning.

It's not really make sense to clone `AlloyDB` or `EthersDB`. `Clone` was also added by me in a previous PR, but I recently found it was a mistake. Generally, it is due to:

1. Both `Provider` and `Middleware` are only safe to clone within the same runtime. If you send it across runtimes (or threads), many internal things and assumptions could break.
2. Even if ensure you are in the same async context all the time or just sync code, it just doesn't make sense to clone the EthersDB because it saves nothing.

To your use case which is similar to mine, my suggestion is:

1. Share a single `Rc<Mutex<CacheDB<EthersDB>>>` across all threads. This avoids duplicate caching for each thread and can speed up your application overall.
2. Clone `CacheDB` members (note they are public) except the `EthersDB` and create a new `EthersDB` instead. This generously is a manual "Clone" implementation.

Reply:
thanks, I have a challenge in using arc<mutex>.

![Uploading image.png…]()

in order to create evm and then transact(), I must do lock. but, about all of my perf. overhead is related to these steps, so it convert my program in a literally single-threaded system.

Reply:
Note `with_db` accepts borrows.

Reply:
you mean `with_ref_db` ?  I want to update the cache in each `transact`